File: src/LowestCommonAncestorofaBinaryTree.java
import java.util.ArrayList;
import java.util.List;

/*
 * The complexity of the algorithm is O(n) where n is no of elements.
 * Space complexity is O(h) where h is height of tree.
 */

public class LowestCommonAncestorofaBinaryTree {

	// without extra space solution.

	public TreeNode lowestCommonAncestor1(TreeNode root, TreeNode p, TreeNode q) {

		if (root == null || root == p || root == q)
			return root;

		TreeNode left = lowestCommonAncestor1(root.left, p, q);
		TreeNode right = lowestCommonAncestor1(root.right, p, q);

		if (left != null && right != null)
			return root;
		else if (left != null)
			return left;
		else if (right != null)
			return right;
		else
			return null;

	}

	// ArrayList solution

	public class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;

		TreeNode() {
		}

		TreeNode(int val) {
			this.val = val;
		}

		TreeNode(int val, TreeNode left, TreeNode right) {
			this.val = val;
			this.left = left;
			this.right = right;
		}
	}

	public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

		List<TreeNode> path1 = new ArrayList<>();
		List<TreeNode> path2 = new ArrayList<>();

		backtrack(root, p, path1);
		backtrack(root, q, path2);

		for (int i = 0; i < path1.size(); i++) {

			if (path1.get(i) != path2.get(i)) {

				return path1.get(i - 1);
			}

		}

		return null;
	}

	private void backtrack(TreeNode root, TreeNode p, List<TreeNode> path) {

		// base

		if (root == null)
			return;

		if (root == p) {
			path.add(root);
			path.add(root);
			return;
		}

		// logic

		path.add(root);

		backtrack(root.left, p, path);
		backtrack(root.right, p, path);

		if (path.get(path.size() - 1) == p)
			return;

		path.remove(path.size() - 1);

	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}

File: Lowest_Common_Ancestor_of_a_Binary_Tree.py
# Time Complexity :O(N)
# Space Complexity :O(H)
# Did this code successfully run on Leetcode : Yes

# Any problem you faced while coding this :

# Your code here along with comments explaining your approach


from collections import deque
import copy as cp


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:

    def helper(self,root, p, q, path):
        #base
        if(root==None):
            return 
        if(root==p):
            self.p_path+=cp.deepcopy(path)
            self.p_path.append(root)
            # return
        if(root==q):
            self.q_path+=cp.deepcopy(path)
            self.q_path.append(root)
            # return/
        #logic
        # action
        path.append(root)
        # resurse
        self.helper(root.left, p, q, path)
        self.helper(root.right, p, q, path)
        # backtrack
        path.pop()


    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # if(root==None or root==p or root==q):
        #     return root
        # l=self.lowestCommonAncestor(root.left, p, q)
        # r=self.lowestCommonAncestor(root.right, p, q)

        # if(l!=None and r!=None):
        #     return root
        # elif(l!=None):
        #     return l
        # elif(r!=None):
        #     return r
        self.p_path=[]
        self.q_path=[]
        self.helper(root, p, q, [])
        # print(self.q_path, self.p_path)
        l=min(len(self.p_path), len(self.q_path))
        # print(self.p_path)
        # print(self.q_path)
        for i in range(len(self.p_path)):
            if(self.p_path[i]!=self.q_path[i]):
                break
        return self.p_path[i-1]


        